pages/papers/attention-is-all-you-need.mdx

```diff
@@ -1,3 +1,223 @@
+
+> The original transformer used a sinusodial embedding which gives a more absolute concept of position. Current work with implementations such as Rotatory Embeddings provide a way for transformers to learn more relative representations of positions in the sequence.
+
+![Positional Encoding](/positional-encoding.png)
+
+Right now our embedding conveys a sense of semantic meaning. But there's no positional information. Let's take the two sentences below to understand what this means.
+
+1. There was a car in front of me, it was red
+2. There was a car in front of me, it was challenging to swerve around it
+
+There are a few things to notice about this sentence
+
+- `it` appears twice - but it refers to different things each time. In the first instance, it refers to the act of driving around the car while the second instance refers to the car.
+- `it` is heavily context dependent, in the first sentence, `it` refers to the car instead!
+
+As a result, we use positional encodings to imbue our model's embeddings with information on the position of each word. The model learns these transformations over time. These encodings are computed once and then frozen in time.
+
+> In the paper, they state that they experimented with learned positional embeddings and found that the two versions produced nearly identical results. 
+
+![Positional Encoding](/positional-encoding.png)
+
+### Scaled-Dot Product Attention and Multi-Head Attention
+
+Attention is a method by which we can compute a new representation of a token given some sort of weighted sum of the representations of the tokens around it. This is done through Scaled Dot Product Attention.
+
+#### Scaled Dot Product Attention
+
+![Transformer Attention](/transformer-attention.png)
+
+Scaled Dot Product Attention is simply an operation by which we can calculate a weighted sum of values
+
+$$
+Attention(Q,K,V) = softmax(\frac{QK^{T}}{\sqrt{d_k}})V
+$$
+
+Note that Q,K and V do not have the be the same matrix. Q and K will have the same dimensions `d_k` and V will have the dimensions of `d_v` . In the paper, the model has an embedding dimension `d_model` of `512` where `d_k = d_v = d_model` .
+
+There are two important things to note here
+
+1. We scale $QK^T$ by $\sqrt{d_k}$ because of numerical instability. This is because the product of two matrices has a significantly greater variance and tends to lead to large values being generated. Scaling down allows for a smaller value to be used when computing the softmax, resulting in better results
+2. The Dot Product is chosen because it is much faster and more space-efficient in practice, since it can be implemented using highly optimised matrix multiplication code.
+
+The model tunes the values of Q and K so that it learns how to assign the proper weightage to each token for a given sequence. We can intuitively understand the Query matrix as a representation of a sequence seeking a match while the Key matrix is a representation of a sequence's meaning.
```

Comment:
Not too sure how to phrase the "seeker" and the has a value representation

Reply:
consider pg 4 of https://web.stanford.edu/class/cs224n/readings/cs224n-self-attention-transformers-2023_draft.pdf
and slide 28 of https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1224/slides/cs224n-2022-lecture09-transformers.pdf

In order to build a contextual representation of a token in the sequence, the model finds out what information in the values (V) are relevant. It does so by comparing the Q against all K and from this comparison, finds the relative importance of each V to build the contextual state.